### PR TITLE
chore: network module for BiDi

### PR DESCRIPTION
**Pull request by Lightning00Blade**

### CODE REVIEW

File: test/src/navigation.spec.ts [hunk 1]
```diff
@@ -144,15 +144,17 @@ describe('navigation', function () {
...
-      let error!: Error;
-      await page.goto('asdfasdf').catch(error_ => {
-        return (error = error_);
-      });
-      if (isChrome) {
-        expect(error.message).toContain('Cannot navigate to invalid URL');
-      } else {
-        expect(error.message).toContain('Invalid url');
-      }
+      try {
```
Comment (OrKoN) [inline test/src/navigation.spec.ts]: why try/catch here?

File: tools/mochaRunner/src/utils.ts [hunk 2]
```diff
@@ -76,14 +76,16 @@ export function printSuggestions(
...
-      'The recommendations are based on the following applied expectaions:'
-    );
-    prettyPrintJSON(
-      toPrint.map(item => {
-        return item.basedOn;
-      })
-    );
+    if (action !== 'remove') {
+      console.log(
+        'The recommendations are based on the following applied expectaions:'
```
Comment (OrKoN) [inline tools/mochaRunner/src/utils.ts]: let's fix a typo in `expectaions`

File: tools/mochaRunner/src/utils.ts [hunk 3]
```diff
@@ -76,14 +76,16 @@ export function printSuggestions(
...
     );
-    console.log(
-      'The recommendations are based on the following applied expectaions:'
-    );
-    prettyPrintJSON(
-      toPrint.map(item => {
-        return item.basedOn;
-      })
-    );
+    if (action !== 'remove') {
```
Comment (OrKoN) [inline tools/mochaRunner/src/utils.ts:79]: why do we want to hide if the action is remove?

File: packages/puppeteer-core/src/common/bidi/FrameManager.ts [hunk 4]
```diff
@@ -60,7 +62,10 @@ export class FrameManager extends EventEmitter {
     this.#page = page;
     this._contextId = info.context;
     this._timeoutSettings = timeoutSettings;
+
+    this._networkManager = new NetworkManager(this.#connection, this);
```
Comment (OrKoN) [inline packages/puppeteer-core/src/common/bidi/FrameManager.ts]: any reason for putting the network manager on the frame manager?

File: packages/puppeteer-core/src/api/HTTPResponse.ts [hunk 5]
```diff
@@ -67,7 +67,9 @@ export class HTTPResponse {
    * True if the response was successful (status in the range 200-299).
    */
   ok(): boolean {
-    throw new Error('Not implemented');
+    // TODO: document === 0 case?
```
Comment (OrKoN) [inline packages/puppeteer-core/src/api/HTTPResponse.ts:70]: is there a left-over common implementation?

File: tools/mochaRunner/src/utils.ts (hunk 3 above)
Comment (Lightning00Blade) [reply tools/mochaRunner/src/utils.ts:79]: If we want to remove Expectation A its based on Expectation A, no need to show it 2 times.

File: packages/puppeteer-core/src/common/bidi/HTTPRequest.ts [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    this.#postData = undefined;
+    this.#initiator = event.initiator;
+    this.#frame = frame;
+
+    this._requestId = event.request.request;
+    this._redirectChain = redirectChain ?? [];
+    this._navigation = event.navigation;
+
+    for (const {name, value} of event.request.headers) {
+      // TODO: How to handle Binary Headers
```
Comment (OrKoN) [inline packages/puppeteer-core/src/common/bidi/HTTPRequest.ts:62]: do we support binary headers? is it a bidi feature or Puppeteer's?

Comment (Lightning00Blade) [reply packages/puppeteer-core/src/common/bidi/HTTPRequest.ts:62]: It's a BiDi feature - https://w3c.github.io/webdriver-bidi/#type-network-Header will leave the link the the comment

File: packages/puppeteer-core/src/api/HTTPResponse.ts (hunk 5 above)
Comment (Lightning00Blade) [reply packages/puppeteer-core/src/api/HTTPResponse.ts:70]: This implementation is common for both  BiDi and CDP `ok()`. Comment is from the CDP file just moved to API.

File: test/src/navigation.spec.ts (hunk 1 above)
Comment (Lightning00Blade) [reply test/src/navigation.spec.ts]: Leftover from testing. Reverting.

File: packages/puppeteer-core/src/api/HTTPResponse.ts (hunk 5 above)
Comment (OrKoN) [reply packages/puppeteer-core/src/api/HTTPResponse.ts:70]: I don't see the override in the subclass being removed with this.

File: packages/puppeteer-core/src/common/bidi/NetworkManager.ts [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+import {FrameManager} from './FrameManager.js';
+import {HTTPRequest} from './HTTPRequest.js';
+import {HTTPResponse} from './HTTPResponse.js';
+
+/**
+ * @internal
+ */
+export class NetworkManager extends EventEmitter {
+  #connection: Connection;
+  #frameManager: FrameManager;
```
Comment (OrKoN) [inline packages/puppeteer-core/src/common/bidi/NetworkManager.ts]: so why do we need frameManager here? can we avoid checking if the context is known?

File: packages/puppeteer-core/src/common/bidi/NetworkManager.ts [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+  #subscribedEvents = new Map<string, Handler<any>>([
+    ['network.beforeRequestSent', this.#onBeforeRequestSent.bind(this)],
+    ['network.responseStarted', this.#onResponseStarted.bind(this)],
+    ['network.responseCompleted', this.#onResponseCompleted.bind(this)],
+    ['network.fetchError', this.#onFetchError.bind(this)],
+  ]) as Map<Bidi.Message.EventNames, Handler>;
+
+  _requestMap = new Map<string, HTTPRequest>();
```
Comment (OrKoN) [inline packages/puppeteer-core/src/common/bidi/NetworkManager.ts]: what are the keys here?

File: packages/puppeteer-core/src/common/bidi/HTTPRequest.ts [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+  ResourceType,
+} from '../../api/HTTPRequest.js';
+
+import {HTTPResponse} from './HTTPResponse.js';
+
+/**
+ * @internal
+ */
+export class HTTPRequest extends BaseHTTPRequest {
+  override _response: HTTPResponse | null = null;
```
Comment (OrKoN) [inline packages/puppeteer-core/src/common/bidi/HTTPRequest.ts:30]: let's replace the next three with proper getters/setters? It does not look like they need to be on the base class as we have specialized network manager implementations?

File: packages/puppeteer-core/src/common/bidi/NetworkManager.ts (hunk 8 above)
Comment (OrKoN) [inline packages/puppeteer-core/src/common/bidi/NetworkManager.ts]: #requestMap

File: packages/puppeteer-core/src/common/bidi/NetworkManager.ts [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        return request.response();
+      }
+    }
+
+    return null;
+  }
+
+  dispose(): void {
+    this.removeAllListeners();
+    this._requestMap = new Map();
```
Comment (OrKoN) [inline packages/puppeteer-core/src/common/bidi/NetworkManager.ts]: this.#requestMap.clear()

**Review by OrKoN — APPROVED** [review]

**Review by OrKoN — APPROVED** [review]

**Pull request merged**